x-pack/platform/plugins/shared/streams_app/public/components/stream_list_view/tree_table.tsx

```diff
@@ -0,0 +1,169 @@
+import { StreamsAppSearchBar } from '../streams_app_search_bar';
+import { DocumentsColumn } from './documents_column';
+import { useStreamsAppRouter } from '../../hooks/use_streams_app_router';
+import { RetentionColumn } from './retention_column';
+
+export function StreamsTreeTable({
+  timefilter,
+  loading,
+  streams,
+}: {
+  timefilter: TimefilterHook; // Workaround to keep state in sync
+  streams: ListStreamDetail[] | undefined;
+  loading?: boolean;
+}) {
+  const router = useStreamsAppRouter();
+  const items = React.useMemo(() => flattenTrees(asTrees(streams ?? [])), [streams]);
+
+  return (
+    <EuiInMemoryTable
+      loading={loading}
+      columns={[
+        {
+          field: 'name',
+          name: i18n.translate('xpack.streams.streamsTreeTable.nameColumnName', {
+            defaultMessage: 'Name',
+          }),
+          width: '40%',
+          render: (name: StreamTreeWithLevel['name'], item) => (
+            <EuiFlexGroup
+              alignItems="center"
+              gutterSize="s"
+              responsive={false}
+              className={css`
+                margin-left: ${item.level * parseInt(euiThemeVars.euiSizeXL, 10)}px;
+              `}
+            >
+              <EuiFlexItem grow={false}>
+                {item.children.length > 0 ? (
+                  <EuiIcon type="arrowDown" color="primary" size="s" />
+                ) : (
+                  <EuiIcon type="empty" color="primary" size="s" />
+                )}
+              </EuiFlexItem>
+              <EuiFlexItem grow={false}>
+                <EuiLink
+                  data-test-subj="streamsAppStreamNodeLink"
+                  href={router.link('/{key}', { path: { key: name } })}
+                >
+                  {name}
+                </EuiLink>
+              </EuiFlexItem>
+            </EuiFlexGroup>
+          ),
+        },
+        {
+          field: 'documents',
+          name: i18n.translate('xpack.streams.streamsTreeTable.documentsColumnName', {
+            defaultMessage: 'Documents',
+          }),
+          width: '40%',
+          render: (_, item) => (
+            <DocumentsColumn timefilter={timefilter} indexPattern={item.name} numDataPoints={25} />
+          ),
+        },
+        {
+          field: 'effective_lifecycle',
+          name: i18n.translate('xpack.streams.streamsTreeTable.retentionColumnName', {
+            defaultMessage: 'Retention',
+          }),
+          width: '20%',
+          render: (_, item) => <RetentionColumn lifecycle={item.effective_lifecycle} />,
+        },
+      ]}
+      itemId="name"
+      items={items}
+      pagination={true}
+      search={{
+        box: {
+          incremental: true,
+        },
+        toolsRight: (
+          <StreamsAppSearchBar
+            onQuerySubmit={({ dateRange }, isUpdate) => {
+              if (dateRange && isUpdate) {
+                timefilter.setTimeRange(dateRange);
+              }
```

Comment:
Should `refreshAbsoluteTimeRange` be in an else branch here?

Reply:
I don't think so. I want the absolute time range to be refreshed irregardless of whether the relative time range changed. For example when a user has `now-15s` selected, opens the time range picker and selects the same range again I think the histograms should refresh based on that selection.

Reply:
`setTimeRange` is implicitly updating the absolute time range as well, right? It updates the timefilter service and that in turn updates the absolute time range here:

https://github.com/elastic/kibana/blob/86a71033b3d81687d8ae3a07e654d12c5d757f17/src/platform/plugins/shared/data/public/query/timefilter/use_timefilter.ts#L37

Not sure whether it matters, just stood out to me.

Reply:
The hook is currently a half way house between both approaches. Consumers need to explicitly refresh the absolute time when some interaction happened but the relative time range did not change (e.g. when selecting the same time range again or when clicking the refresh button or enabling auto-refresh). Consumers do not need to do this however when selecting a different time range. In that case the absolute time is refreshed automatically. This is fairly confusing. the behaviour should be consistent. I think the easiest approach from a consumer perspective would be to handle each of these cases for the user implicitly.

Reply:
I went ahead and created a hook with context provider that solves most of the issues with the current implementation. Might make sense for the streams app search bar component to integrate with that natively so that we don't need to manually set the callback methods everywhere but we can do that incrementally when all the different use cases are clearer.